### PR TITLE
fix(auth): use sha-512 for auth tokens

### PR DESCRIPTION
**Pull request by davidmehren**
### Component/Part
Auth Tokens

### Description
Bcrypt hashes are too slow to be validated on every request.
As our tokens are random and have a fixed length, it is reasonable
to use SHA-512 instead.

SHA-512 is recommended by the BSI:
https://www.bsi.bund.de/SharedDocs/Downloads/EN/BSI/Publications/TechGuidelines/TG02102/BSI-TR-02102-1.pdf?__blob=publicationFile

Fixes https://github.com/hedgedoc/hedgedoc/issues/1881

### Steps

- [x] Added implementation
- [x] Added / updated tests
- [x] I read the [contribution documentation](https://github.com/hedgedoc/hedgedoc/blob/develop/CONTRIBUTING.md) and
  made sure that:
  - My commits are signed-off to accept the DCO.
  - This PR targets the correct branch: `master` for 1.x & docs, `develop` for 2.x

### CODE REVIEW

File: src/auth/auth.service.ts [hunk 1]
```diff
@@ -70,9 +66,12 @@ export class AuthService {
         `User '${user.username}' has already 200 tokens and can't have anymore`,
       );
     }
-    const secret = bufferToBase64Url(randomBytes(54));
+    const secret = bufferToBase64Url(randomBytes(64));
```
Comment (DerMolly) [inline src/auth/auth.service.ts:69]: This change is currently a bit unexplained. I'd suggest changing that.

**Review by DerMolly — APPROVED** [review]
LGTM

**Comment by SISheogorath** [comment]
> SHA-512 is recommended by the BSI:
https://www.bsi.bund.de/SharedDocs/Downloads/EN/BSI/Publications/TechGuidelines/TG02102/BSI-TR-02102-1.pdf?__blob=publicationFile

Uhm, which section are you refering to? I don't think they talk about our use-case anywhere in that document. Section 6.1 recommends to do challenge-response for authentication with a shared secret. Otherwise it's quite silent on this part.

File: src/auth/auth.service.ts [hunk 2]
```diff
@@ -70,9 +66,12 @@ export class AuthService {
...
     const keyId = bufferToBase64Url(randomBytes(8));
-    const accessToken = await hashPassword(secret);
+    const accessToken = crypto
```
Comment (InnayTool) [inline src/auth/auth.service.ts:71]: I think that it should be named `accessTokenHash` to make clearer it is not the actual token.

**Review by InnayTool — CHANGES_REQUESTED** [review]

**Comment by davidmehren** [comment]
I updated the PR according to the reviews and expanded some comments

**Comment by davidmehren** [comment]
Regarding the use of SHA-512:
Unfortunately, I could not find any explicit documentation about our exact use-case.
Most docs describe classic password-saving scenarios and recommend bcrypt, scrypt or argon2. These hashing functions are slow to stop brute-force or dictionary attacks, which would expose the original, user-provided password, which might have been reused across multiple services.

We have a very different scenario: Our API tokens are 64 bytes of cryptographically strong pseudorandom data. Brute-force or dictionary attacks are therefore virtually impossible, and they are not reused across multiple services. We therefore need to only guard against one scenario: An attacker gains read-only access to the database. Saving only hashes in the database prevents the attacker from authenticating themselves as a user. The hash-function does not need to be very slow, as the randomness of the original token prevents inverting the hash. The function actually needs to be reasonably fast, as the hash must be computed on every request to the public API. SHA-512 (or alternatively SHA3) fits this use-case.

**Comment by SISheogorath** [comment]
@davidmehren 👍🏻 This sounds better :)

**Comment by davidmehren** [comment]
@SISheogorath Great idea, I added the explanation to the docs.

File: docs/content/dev/2.0.md [hunk 3]
```diff
@@ -39,3 +39,28 @@ Because we need to have empty constructors in our entity classes for TypeORM to
 - Should either return a complete and fully useable instance or return a Pick/Omit type.
 - Exceptions to these rules are allowed, if they are mentioned in the method documentation
 
+## Auth tokens for the public API
+The public uses bearer tokens for authentication.
```
Comment (InnayTool) [inline docs/content/dev/2.0.md:43]:
```suggestion
The public API uses bearer tokens for authentication.
```

**Review by InnayTool — CHANGES_REQUESTED** [review]

**Review by InnayTool — APPROVED** [review]

**Pull request merged**